src/psyclone/psyGen.py

```diff
@@ -4465,6 +4639,201 @@ def _number_handler(self, node, parent):
+
+    def __str__(self):
+        return (self.name + "<" + self.datatype + ", " + str(self.shape) +
+                ", " + self.access + ">")
+
+
+class SymbolTable(object):
+    '''
+    Encapsulates the symbol table and provides methods to declare new symbols
+    and look up existing symbols. It is implemented as a single scope
+    symbol table (nested scopes not supported).
+    '''
+    def __init__(self):
+        # Dict of Symbol objects with the symbol names as keys.
+        self._symbols = {}
+        # Ordered list of the arguments.
+        self._argument_list = []
+
+    def declare(self, name, datatype, shape, access):
+        '''
+        Declare a new symbol in the symbol table.
+
+        :param str name: Name of the symbol.
+        :param str datatype: Datatype of the symbol.
+        :param shape: Shape of the symbol (empty list represents an scalar
+                      symbol).
+        :type shape: list of integers
+        :param str access: Information about the variable declaration
+                           attributes.
+        '''
+        if name in self._symbols:
+            raise InternalError("Symbol table already contains a symbol with"
+                                " name '{0}'.".format(name))
+        self._symbols[name] = Symbol(name, datatype, shape, access)
+
+    def specify_argument_list(self, argument_name_list):
+        '''
+        Keep track of the arguments order and provide the argument access
+        attribute if it was not available on the variable declaration.
+
+        :param list argument_name_list: Ordered list of the argument names.
+        '''
+        for name in argument_name_list:
+            symbol = self.lookup(name)
+            if symbol.access == 'local':
+                symbol.access = 'readwrite_arg'
+            self._argument_list.append(symbol)
+
+    def lookup(self, name):
+        '''
+        Look up a symbol in the symbol table.
+
+        :param str name: Name of the symbol
+        :raises KeyError: If the given name is not in the Symbol Table.
+        '''
+        return self._symbols[name]
+
+    def view(self):
+        '''
+        Print a representation of this Symbol Table to stdout.
+        '''
+        print("Symbol Table:")
+        for symbol in self._symbols.values():
+            print(str(symbol))
+
+    def __str__(self):
+        return ("Symbol Table:\n" +
+                "\n".join(map(str, self._symbols.values())) +
+                "\n")
+
+
+class KernelSchedule(Schedule):
+    '''
+    A kernelSchedule inherits the functionality from Schedule and adds a symbol
+    table to keep a record of the declared variables and their attributes.
+
+    :param str name: Kernel subroutine name
+    :type name: string
+    '''
+
+    def __init__(self, name):
+        super(KernelSchedule, self).__init__(None, None)
+        self._name = name
+        self._symbol_table = SymbolTable()
+
+    @property
+    def symbol_table(self):
```

Comment:
Please document this method (i.e. the return type).

Reply:
Done